File: arb-www/src/schema/rollingInsights/leagueResponse.ts
import { BaseDTO } from '../BaseDTO';
import type { NBARollingInsightsBoxScoreGame } from './nba/boxScore';
import type { NBARollingInsightsPlayerProfile } from './nba/playerProfile';
import type { NBARollingInsightsScheduleGame } from './nba/schedule';
import type { NBARollingInsightsTeamProfile } from './nba/teamProfile';

/**
 * Rolling Insights DataType enum
 */
export enum RollingInsightsDataType {
  Schedule = 'schedule',
  CurrentGames = 'current_games',
  TeamProfiles = 'team_profiles',
  Headshots = 'headshots',
  Stadiums = 'stadiums',
  Standings = 'standings',
  BoxScore = 'box_score',
  PlayByPlay = 'play_by_play',
  Odds = 'odds',
  GameByDate = 'game_by_date',
  Scores = 'scores',
}

/**
 * Rolling Insights League Response
 */
export class RollingInsightsLeagueResponse extends BaseDTO {
  league!: string;
  data_type!: RollingInsightsDataType;
  data!: RollingInsightsLeagueData;
  filtered_count!: number;
  total_count!: number;
}

/**
 * Rolling Insights League Data union type
 */
export type RollingInsightsLeagueData = 
  | { Nba: NBARollingInsightsData };

/**
 * Rolling Insights NBA Data
 */
export type NBARollingInsightsData = 
  | { Schedule: NBARollingInsightsScheduleGame[] }
  | { CurrentGames: NBARollingInsightsScheduleGame[] }
  | { TeamProfiles: NBARollingInsightsTeamProfile[] }
  | { PlayerProfiles: NBARollingInsightsPlayerProfile[] }
  | { BoxScore: NBARollingInsightsBoxScoreGame[] };

File: arb-rs/src/schema/rolling_insights/league_response.rs
use serde::{Deserialize, Serialize};

use super::{
    data_type::DataType,
    nba::{
        box_score::NBARollingInsightsBoxScoreGame,
        player_profile::NBARollingInsightsPlayerProfile,
        schedule::NBARollingInsightsScheduleGame,
        team_profile::NBARollingInsightsTeamProfile,
    },
};

/// Generic response wrapper for different leagues from Rolling Insights
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RollingInsightsLeagueResponse {
    pub league: String,
    pub data_type: DataType,
    pub data: RollingInsightsLeagueData,
    pub filtered_count: usize,
    pub total_count: usize,
}

/// League-specific data types from Rolling Insights
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RollingInsightsLeagueData {
    Nba(Box<NBARollingInsightsData>),
}

/// NBA-specific data types from Rolling Insights
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum NBARollingInsightsData {
    Schedule(Vec<NBARollingInsightsScheduleGame>),
    CurrentGames(Vec<NBARollingInsightsScheduleGame>),
    TeamProfiles(Vec<NBARollingInsightsTeamProfile>),
    PlayerProfiles(Vec<NBARollingInsightsPlayerProfile>),
    BoxScore(Vec<NBARollingInsightsBoxScoreGame>),
}
